server/src/main/java/org/elasticsearch/index/shard/IndexWriteLoadForecast.java

```diff
@@ -0,0 +1,162 @@
+        this(in.readDoubleArray());
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeDoubleArray(forecastedShardWriteLoad);
+    }
+
+    public static ClusterState maybeIncludeWriteLoadForecast(
+        IndexAbstraction.DataStream dataStream,
+        ClusterState clusterState,
+        TimeValue maxIndexAge,
+        TimeValue minShardUptime
+    ) {
+        final IndexMetadata writeIndex = clusterState.metadata().getIndexSafe(dataStream.getWriteIndex());
+        if (IndexSettings.FORECAST_WRITE_LOAD_SETTING.get(writeIndex.getSettings()) == false) {
+            return clusterState;
+        }
+
+        final int numberOfShards = writeIndex.getNumberOfShards();
+        final double[] totalWriteLoad = new double[numberOfShards];
+        final double[] maxWriteLoad = new double[numberOfShards];
+        final int[] numberOfIndicesTookIntoAccount = new int[numberOfShards];
+
+        for (Index index : dataStream.getIndices()) {
+            final IndexMetadata indexMetadata = clusterState.metadata().getIndexSafe(index);
+            final long indexAge = System.currentTimeMillis() - indexMetadata.getCreationDate();
+            if (index.equals(dataStream.getWriteIndex()) || indexAge > maxIndexAge.millis()) {
+                continue;
+            }
+
+            final IndexWriteLoad writeLoad = indexMetadata.getWriteLoad();
+            for (int shardId = 0; shardId < numberOfShards; shardId++) {
+                OptionalDouble writeLoadForShard = writeLoad.getWriteLoadForShard(shardId);
+                OptionalLong uptimeInMillisForShard = writeLoad.getUptimeInMillisForShard(shardId);
+                if (writeLoadForShard.isPresent()) {
+                    assert uptimeInMillisForShard.isPresent();
+                    double shardWriteLoad = writeLoadForShard.getAsDouble();
+                    long shardUptimeInMillis = uptimeInMillisForShard.getAsLong();
+                    if (shardUptimeInMillis > minShardUptime.millis()) {
+                        totalWriteLoad[shardId] += shardWriteLoad;
+                        maxWriteLoad[shardId] = Math.max(shardWriteLoad, maxWriteLoad[shardId]);
+                        numberOfIndicesTookIntoAccount[shardId] += 1;
+                    }
+                }
+            }
+        }
+
+        IndexWriteLoadForecast.Builder projectedWriteLoad = IndexWriteLoadForecast.builder(numberOfShards);
+        boolean modified = false;
+        for (int shardId = 0; shardId < totalWriteLoad.length; shardId++) {
+            int shardIndicesTookIntoAccount = numberOfIndicesTookIntoAccount[shardId];
+            if (shardIndicesTookIntoAccount > 0) {
+                modified = true;
+                double normalizedShardLoad = (totalWriteLoad[shardId] / shardIndicesTookIntoAccount) / maxWriteLoad[shardId];
+                projectedWriteLoad.withForecastedWriteLoad(shardId, normalizedShardLoad);
```

Comment:
I think we should instead just end up with one load number on the index. This can then be evenly spread across the shards. Trying to preserve any shard stickyness (i.e., routing use) seems heroic and makes it more difficult to disregard shards with short uptimes.